src/data/arena-tag.ts

```diff
@@ -919,6 +919,65 @@ class SafeguardTag extends ArenaTag {
   }
 }
 
+/**
+ * This arena tag facilitates the application of the move Imprison
+ * Imprison remains in effect as long as the source Pokemon is active and present on the field.
+ * Imprison will apply to any opposing Pokemon that switch onto the field as well.
+ */
+class ImprisonTag extends ArenaTrapTag {
+  private source: Pokemon;
+
+  constructor(sourceId: integer, side: ArenaTagSide) {
+    super(ArenaTagType.IMPRISON, Moves.IMPRISON, sourceId, side, 1);
+  }
+
+  /**
+   * This function applies the effects of Imprison to the opposing Pokemon already present on the field.
+   * @param arena
+   */
+  override onAdd(arena: Arena) {
+    this.source = arena.scene.getPokemonById(this.sourceId!)!;
+    if (this.source) {
+      const party = !this.source.isPlayer() ? arena.scene.getPlayerField() : arena.scene.getEnemyField();
+      party?.forEach((p: PlayerPokemon | EnemyPokemon ) => {
+        p.addTag(BattlerTagType.IMPRISON, 1, Moves.IMPRISON, this.sourceId);
+      });
+      arena.scene.queueMessage(i18next.t("battlerTags:imprisonOnAdd", {pokemonNameWithAffix: getPokemonNameWithAffix(this.source)}));
+    }
+  }
```

Comment:
NIT:
use destructuring to your advantage
```suggestion
  override onAdd({ scene }: Arena) {
    this.source = scene.getPokemonById(this.sourceId!)!;
    if (this.source) {
      const party = !this.source.isPlayer() ? scene.getPlayerField() : scene.getEnemyField();
      party?.forEach((p: PlayerPokemon | EnemyPokemon ) => {
        p.addTag(BattlerTagType.IMPRISON, 1, Moves.IMPRISON, this.sourceId);
      });
      scene.queueMessage(i18next.t("battlerTags:imprisonOnAdd", {pokemonNameWithAffix: getPokemonNameWithAffix(this.source)}));
    }
  }
```